File: Friends Pairing Problem.py
class Solution:
    def countFriendsPairings(self, n):
        # code here 
        mod = 10**9+7
        dp = [0]*(n+1)
        dp[0] = 1
        dp[1] = 1
        for i in range(2,n+1) :
            dp[i] = (dp[i-1]+(i-1)*dp[i-2])%mod
            
        return dp[n]

File: Friends Pairing Problem.cpp
@easy c++ sol
class Solution
{
public:
//int mod = 1000000007;
        
    long long  solve(int n , vector<int>& dp)
    {  int mod = 1000000007;
        
        if(n==1)
        return 1;
        if(n==2)
        return 2;
        if(dp[n]!=-1)
        return dp[n];
        
        int r = solve(n-1, dp)*1%mod + ((n-1)*(solve(n-2, dp)%mod))%mod;
        return  dp[n]=r%mod;
        
    }
    int countFriendsPairings(int n) 
    { 
        vector<int>dp(n+1, -1);
        int x= solve(n , dp);
        return  x;
    }
};
